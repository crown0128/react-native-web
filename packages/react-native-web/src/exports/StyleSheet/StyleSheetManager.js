/**
 * Copyright (c) 2016-present, Nicolas Gallagher.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import createAtomicRules from './createAtomicRules';
import hash from '../../vendor/hash';
import initialRules from './initialRules';
import WebStyleSheet from './WebStyleSheet';

const emptyObject = {};
const STYLE_ELEMENT_ID = 'react-native-stylesheet';

const createClassName = (prop, value) => {
  const hashed = hash(prop + value);
  return process.env.NODE_ENV !== 'production' ? `rn-${prop}-${hashed}` : `rn-${hashed}`;
};

export default class StyleSheetManager {
  _cache = {
    byClassName: {},
    byProp: {}
  };

  constructor(id) {
    this._id = this._sheet = new WebStyleSheet(STYLE_ELEMENT_ID);
    initialRules.forEach(rule => {
      this._sheet.insertRuleOnce(rule);
    });
  }

  getClassName(prop, value) {
    const cache = this._cache.byProp;
    return cache[prop] && cache[prop].hasOwnProperty(value) && cache[prop][value];
  }

  getDeclaration(className) {
    const cache = this._cache.byClassName;
    return cache[className] || emptyObject;
  }

  getStyleSheets() {
    const { cssText } = this._sheet;

    return [
      {
        id: STYLE_ELEMENT_ID,
        textContent: cssText
      }
    ];
  }

  injectDeclaration(prop, value): string {
    let className = this.getClassName(prop, value);
    if (!className) {
      className = createClassName(prop, value);
      this._addToCache(className, prop, value);
      const rules = createAtomicRules(`.${className}`, prop, value);
      rules.forEach(rule => {
        this._sheet.insertRuleOnce(rule);
      });
    }
    return className;
  }

  _addToCache(className, prop, value) {
    const cache = this._cache;
    if (!cache.byProp[prop]) {
      cache.byProp[prop] = {};
    }
    cache.byProp[prop][value] = className;
    cache.byClassName[className] = { prop, value };
  }
}
