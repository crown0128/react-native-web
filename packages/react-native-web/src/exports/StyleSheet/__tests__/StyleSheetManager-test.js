/* eslint-env jasmine, jest */

import StyleSheetManager from '../StyleSheetManager';

let styleSheetManager;

describe('apis/StyleSheet/StyleSheetManager', () => {
  beforeEach(() => {
    styleSheetManager = new StyleSheetManager();
  });

  test('getClassName', () => {
    expect(styleSheetManager.getClassName('pointerEvents', 'box-only')).toMatchSnapshot();
    const className = styleSheetManager.setDeclaration('width', '100px');
    expect(styleSheetManager.getClassName('width', '100px')).toEqual(className);
  });

  test('getDeclaration', () => {
    const className = styleSheetManager.setDeclaration('width', '100px');
    expect(styleSheetManager.getDeclaration(className)).toEqual({
      prop: 'width',
      value: '100px'
    });
  });

  test('getStyleSheets', () => {
    styleSheetManager.setDeclaration('--test-property', 'test-value');
    expect(styleSheetManager.getStyleSheets()).toMatchSnapshot();
  });

  test('setDeclaration', () => {
    styleSheetManager._webStyleSheet.insertRuleOnce = (rule, position) => {
      // check for regressions in CSS write path (e.g., 0 => 0px)
      expect(rule.indexOf('-webkit-flex-shrink:0;')).not.toEqual(-1);
    };
    styleSheetManager.setDeclaration('flexShrink', 0);
  });
});
