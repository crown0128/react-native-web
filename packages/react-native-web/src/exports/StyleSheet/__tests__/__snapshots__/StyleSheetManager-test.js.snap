// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apis/StyleSheet/StyleSheetManager getClassName 1`] = `"rn-pointerEvents-ah5dr5"`;

exports[`apis/StyleSheet/StyleSheetManager getStyleSheets 1`] = `
Array [
  Object {
    "id": "react-native-stylesheet",
    "textContent": "html{-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:rgba(0,0,0,0);}
body{margin:0;}
@media all{button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0;}}
@media all{input::-webkit-inner-spin-button,input::-webkit-outer-spin-button,input::-webkit-search-cancel-button,input::-webkit-search-decoration,input::-webkit-search-results-button,input::-webkit-search-results-decoration{display:none;}}
@keyframes rn-ActivityIndicator-animation{0%{-webkit-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);transform:rotate(360deg);}}
@keyframes rn-ProgressBar-animation{0%{-webkit-transform:translateX(-100%);transform:translateX(-100%);}100%{-webkit-transform:translateX(400%);transform:translateX(400%);}}
.rn-pointerEvents-ah5dr5 > *{pointer-events:none}
.rn-pointerEvents-ah5dr5{pointer-events:auto !important}
.rn-pointerEvents-12vffkv > *{pointer-events:auto}
.rn-pointerEvents-12vffkv{pointer-events:none !important}
.rn-pointerEvents-105ug2t{pointer-events:auto !important}
.rn-pointerEvents-633pao{pointer-events:none !important}
.rn---test-property-ax3bxi{--test-property:test-value}",
  },
]
`;
